dask/dataframe/shuffle.py

```diff
@@ -183,6 +183,11 @@ def set_index(
     """See _Frame.set_index for docstring"""
     if isinstance(index, Series) and index._name == df.index._name:
         return df
+
+    if is_index_like(index):
```

Comment:
This will catch both pandas Indexes and dask Indexes. I'm not sure if pandas indexes even need to be converted to a Series or not? And for dask Indexes, I'm not sure if the logic here is sufficient.

I don't see anything in the current code that uses divisions on `index` if they're known and `divisions` was passed as None.

Note that in the original issue, I was discussing changing the logic on the `DataFrame.set_index` method, not the plain `set_index` function. It's _maybe_ sensible for top-level `set_index` to handle all this (really, it should probably be renamed to `set_index_shuffle` or something), but that would involve moving logic that's currently in `DataFrame.set_index` into there, or elsewhere. I think there's an unwritten assumption in `shuffle.set_index` that the `index` is unsorted.

Reply:
Put another way: I'm not sure if passing a dask Series with known divisions into `DataFrame.set_index` takes a `set_sorted_index` fastpath right now either. But it should. Getting that to work correctly would be a good first step; then extending that logic to also handle a dask Index is trivial.